### ISSUE TITLE
Normalize DOI, ISBN and PMID in PublicationIdentifiers and PubHash data

### ISSUE THREAD

**Issue by peetucket** [sul-dlss/sul_pub#241]
Our publication identifiers database has inconsistent use of DOI formatting, e.g. ```["10.1016/j.tecto.2012.04.002", "10.1155/2013/342479", "10.1001/jama.2011.1627", "10.1016/j.jcrs.2011.05.043.", "10.1007/s11426-009-0030-2", "DOI: 10.1002/hbm.21338", "10.1029/2005TC001887", "doi44-22/1034"]```

**Comment by darrenleeweber** [comment]
Collect examples of things to normalize:
- [ ] `10.1016/j.jpsychires.2013.07.004. [Epub ahead of print]`

**Comment by darrenleeweber** [comment]
Maybe we can use this gem - https://github.com/altmetric/identifiers

**Comment by darrenleeweber** [comment]
The log file of running the Altmetrics identifier script in #246 on a prod-db dump is too large to upload into github (> 10Mb)

**Comment by darrenleeweber** [comment]
Some example errors identified by the altmetric identifier script:
```
$ grep 'ERROR' log/altmetric_parser.log | grep -v 'BLANK' | sed -e 's/.*ERROR -- : //g' | head -n50
PublicationIdentifier 889213: SulPubID 242841: '' extracted from 'DOI999999'
PublicationIdentifier 889498: SulPubID 242934: '' extracted from '10.1378'
PublicationIdentifier 889500: SulPubID 242935: '' extracted from '10.1002'
PublicationIdentifier 889585: SulPubID 242989: '' extracted from '10.1523'
PublicationIdentifier 889733: SulPubID 243041: '' extracted from '978-1-137-03424-3'
PublicationIdentifier 889735: SulPubID 243042: '' extracted from '978-1-137-03403-8'
PublicationIdentifier 889737: SulPubID 243043: '' extracted from '978-1-137-03420-5'
PublicationIdentifier 889739: SulPubID 243044: '' extracted from '978-1-137-03428-1'
PublicationIdentifier 890388: SulPubID 243439: '' extracted from 'DOI888-chg'
PublicationIdentifier 890389: SulPubID 243440: '' extracted from 'DOI8888-chg'
PublicationIdentifier 890474: SulPubID 243496: '' extracted from 'DOI222-30/4578'
PublicationIdentifier 890475: SulPubID 243497: '' extracted from 'DOI333-10/4567'
PublicationIdentifier 890512: SulPubID 243513: '' extracted from 'doi44-22/1034'
PublicationIdentifier 890523: SulPubID 243519: '' extracted from '3'
PublicationIdentifier 890683: SulPubID 243679: '' extracted from 'S0003-4975'
PublicationIdentifier 890740: SulPubID 243721: '' extracted from 'http://dx.doi.org/null'
PublicationIdentifier 899569: SulPubID 247658: '' extracted from 'e1002850'
PublicationIdentifier 899612: SulPubID 243367: '' extracted from 'http://dx.doi.org/null'
PublicationIdentifier 899778: SulPubID 247933: '' extracted from 'http://www.ncbi.nlm.nih.gov/pubmed/23977410'
PublicationIdentifier 900638: SulPubID 248368: '' extracted from 'http://dx.doi.org/null'
PublicationIdentifier 913651: SulPubID 254017: '' extracted from 'http://dx.doi.org/null'
PublicationIdentifier 913667: SulPubID 254025: '' extracted from 'DOI'
PublicationIdentifier 914716: SulPubID 254447: '' extracted from 'arXiv:1104.4568'
PublicationIdentifier 915055: SulPubID 254606: '' extracted from 'http://dx.doi.org/s00477-012-0'
PublicationIdentifier 915170: SulPubID 255414: '' extracted from 'http://dx.doi.org/www.biogeosciences.net/6/3131/2009'
PublicationIdentifier 915194: SulPubID 255517: '' extracted from 'http://dx.doi.org/104504-1-10'
PublicationIdentifier 915219: SulPubID 255863: '' extracted from 'http://dx.doi.org/101029/2010WR010120'
PublicationIdentifier 915240: SulPubID 255969: '' extracted from 'http://dx.doi.org/1029/2011JC007726'
PublicationIdentifier 915242: SulPubID 255977: '' extracted from 'http://dx.doi.org/10.111/j.1529-8817.2011.01098.x'
PublicationIdentifier 915249: SulPubID 255985: '' extracted from 'http://dx.doi.org/10.1029.2009GB003737'
PublicationIdentifier 915273: SulPubID 256020: '' extracted from 'http://dx.doi.org/1029/2001JC001226'
PublicationIdentifier 915277: SulPubID 256027: '' extracted from 'http://dx.doi.org/10.1029 / 2001GB001444'
PublicationIdentifier 915289: SulPubID 256383: '' extracted from 'http://dx.doi.org//j.advwatres.2012.03.012'
PublicationIdentifier 915303: SulPubID 256508: '' extracted from 'http://dx.doi.org/dx.doi.org//es103761x'
PublicationIdentifier 915323: SulPubID 256691: '' extracted from 'http://dx.doi.org/2010.2051333'
PublicationIdentifier 915361: SulPubID 256782: '' extracted from 'http://dx.doi.org/464'
PublicationIdentifier 915491: SulPubID 257888: '' extracted from 'http://dx.doi.org/1029/2005JB003812'
PublicationIdentifier 915524: SulPubID 257929: '' extracted from 'http://dx.doi.org/10.129/2005JB004081'
PublicationIdentifier 915525: SulPubID 257930: '' extracted from 'http://dx.doi.org/0.1029/170GM01'
PublicationIdentifier 915526: SulPubID 257931: '' extracted from 'http://dx.doi.org/0.1029/170GM01'
PublicationIdentifier 915530: SulPubID 257939: '' extracted from 'http://dx.doi.org/10.129/2004GL020557'
PublicationIdentifier 915623: SulPubID 258286: '' extracted from 'http://dx.doi.org/15,183-15,202'
PublicationIdentifier 915646: SulPubID 258547: '' extracted from 'http://dx.doi.org/10.115/1.1924398'
PublicationIdentifier 915647: SulPubID 258551: '' extracted from 'http://dx.doi.org/10.1016, j.petrol.2004.08.003'
PublicationIdentifier 915651: SulPubID 258618: '' extracted from 'http://dx.doi.org/1016/j.agrformet.2010.12.002'
PublicationIdentifier 915658: SulPubID 258639: '' extracted from 'http://dx.doi.org/1007/s10584-007-9374-9'
PublicationIdentifier 915682: SulPubID 258770: '' extracted from 'http://dx.doi.org/21:GB1018'
PublicationIdentifier 915690: SulPubID 259032: '' extracted from 'http://dx.doi.org/1017/S0016756812000453'
PublicationIdentifier 915692: SulPubID 259034: '' extracted from 'http://dx.doi.org/1007/s11004-012-9407-4'
PublicationIdentifier 915698: SulPubID 259040: '' extracted from 'http://dx.doi.org/10.1016/ j.jsg.2011.09.003.'
```

**Comment by peetucket** [comment]
Is the log file on your laptop?   We could tuck it on the sul-pub-prod server somewhere for now.

How long did the run take?  While it would be nice to deal with this during this workcycle (presumably by looking up these publications by WoS or PMID and seeing if there is an actual DOI we can use and if not just blanking them out), I'm not sure it needs to be done right now.  The reason I ask how long it took to run is maybe we'd run again just before we tried to fix them.

**Comment by darrenleeweber** [comment]
It runs quickly.

**Comment by darrenleeweber** [comment]
Expanding the scope of this to handle DOI, ISBN and PMID, because the Identifier gem supports normalizing these values too and the code is now available in the associated PR #246

**Comment by darrenleeweber** [comment]
Work on #246 is done.   Running final tests on a prod-db dump on my laptop.  Seems to behave as expected so far.
